frontend/src/stimulus/controllers/pattern-autocompleter.controller.ts

```diff
@@ -0,0 +1,308 @@
+      selection?.removeAllRanges();
+      selection?.addRange(postRange);
+
+      this.updateFormInputValue();
+      this.setRange();
+
+      // clear suggestions
+      this.clearSuggestionsFilter();
+      this.hide(this.suggestionsTarget);
+    } else {
+      this.contentTarget.appendChild(tokenElement);
+    }
+  }
+
+  private currentWord():string|null {
+    const selection = document.getSelection();
+    if (selection) {
+      return (selection.anchorNode?.textContent?.slice(0, selection.anchorOffset)
+        .split(' ')
+        .pop() as string)
+        .toLowerCase();
+    }
+
+    return null;
+  }
+
+  private clearSuggestionsFilter():void {
+    const suggestionElements = this.suggestionsTarget.children;
+    for (let i = 0; i < suggestionElements.length; i+=1) {
+      this.show(suggestionElements[i] as HTMLElement);
+    }
+  }
+
+  private filterSuggestions(word:string):void {
+    const suggestionElements = this.suggestionsTarget.children;
+    for (let i = 0; i < suggestionElements.length; i+=1) {
+      const suggestionElement = suggestionElements[i] as HTMLElement;
+      if (!suggestionElement.dataset.value) { continue; }
+
+      if (suggestionElement.textContent?.trim().toLowerCase().includes(word) || suggestionElement.dataset.value.includes(word)) {
+        this.show(suggestionElement);
+      } else {
+        this.hide(suggestionElement);
+      }
+    }
+
+    // show autocomplete
+    this.show(this.suggestionsTarget);
+  }
+
+  private selectSuggestionAt(index:number):void {
+    if (this.selectedSuggestion.element) {
+      this.selectedSuggestion.element.classList.remove('selected');
+      this.selectedSuggestion.element = null;
+    }
+
+    const possibleTargets = this.suggestionsTarget.querySelectorAll('[data-role="suggestion-item"]:not([hidden])');
+    if (possibleTargets.length > 0) {
+      if (index < 0) { index += possibleTargets.length; }
+      index %= possibleTargets.length;
+      const element = possibleTargets[index];
+      element.classList.add('selected');
+      this.selectedSuggestion.element = element as HTMLElement;
+      this.selectedSuggestion.index = index;
+    }
+  }
+
+  private hide(el:HTMLElement):void {
+    el.setAttribute('hidden', 'hidden');
+  }
+
+  private show(el:HTMLElement):void {
+    el.removeAttribute('hidden');
+  }
+
+  private createToken(value:string):HTMLElement {
+    const target = this.tokenTemplateTarget.content?.cloneNode(true) as HTMLElement;
+    const contentElement = target.firstElementChild as HTMLElement;
+    (contentElement.querySelector('[data-role="token-text"]') as HTMLElement).innerText = value;
+    return contentElement;
+  }
+
+  private toHtml(blueprint:string):string {
+    let htmlValue = blueprint.replace(/{{([0-9A-Za-z_]+)}}/g, (_, token:string) => this.createToken(token).outerHTML);
+    if (htmlValue.startsWith('<')) { htmlValue = ` ${htmlValue}`; }
+    if (htmlValue.endsWith('>')) { htmlValue = `${htmlValue} `; }
```

Comment:
These two lines seem to be duplicated with `ensureSpacesAround`. Can we make sure to only implement this once?

E.g. the caller of `toHtml(...)` could then call `ensureSpacesAround()` afterwards. I.e. what I'd try to achieve is to "localize" the workaround, making sure that it's not too far spread out.